### ISSUE TITLE
Topic broadcast support in PubSub subscription

### ISSUE THREAD

**Issue by artursouza** [dapr/dapr#5647]
## In what area(s)?

/area runtime

> /area operator

> /area placement

> /area docs

> /area test-and-release

## Describe the feature
When an application registers subscriptions, the `consumerID` is generated by the component and is shared across all subscriptions. It does not allow a single application to have multiple handles for the same topic (broadcast). This proposal is to allow the consumerID to be overwritten at each subscription via an optional suffix (`subscriptionID`), enabling an app to subscribe to the same topic multiple times. The runtime (components-contrib) would then `subscriptionID` (when present) to compose the final `consumerID` for each component.

Components-contrib: Each PubSub component would need to parse `subscriptionID` when composing the `consumerID` - the code can probably be reused but components would need to be modified to invoke this method. 
SDKs: This would also require SDKs to allow `subscriptionID` to be specified at the subscription's annotation (`WithTopic`, `Topic`, etc).
CRD: This change should also take place in the `Subscription` CRD.

Example of the subscription JSON payload after this change:
```json
[
  {
    "pubsubname": "pubsub",
    "topic": "newOrder",
    "route": "/record",
    "subscriptionID": "record-keeping",
    "
  },
  {
    "pubsubname": "pubsub",
    "topic": "newOrder",
    "route": "/notify",
    "subscriptionID": "notification",
  },
]
```

## Release Note

RELEASE NOTE: Topic broadcast support in PubSub subscription

**Comment by artursouza** [comment]
Not sure if this is worth a proposal in proposals repo. Thoughts?

**Comment by mukundansundar** [comment]
Probably add it to the proposals repo since this does affect multiple repos.

**Comment by yaron2** [comment]
Actually I think this really shouldn't need to go into the proposals repo.

**Comment by artursouza** [comment]
> @artursouza can't this be done only at `daprd` by allowing to override `consumerID`?

I thought about that but `consumerID` seems to be a composite key at this point and expecting the developer to know how to put that in the app in a way that does not conflict with other apps is not idea. Or worse, it will make running with multiple app instances a pain. What about `{podName}`? Would the developer need to put that in his code as well? So, I think we should handle `consumerID` gracefully on behalf of the user. So, I see the `subscriptionID` as a suffix to whatever is the `consumerID` for a component.

**Comment by ItalyPaleAle** [comment]
Tricky thing with this proposal is that it's very much broker-dependent. Related issue: https://github.com/dapr/dapr/issues/5244

As it is today, we have 3 kinds of brokers:

1. Some do fanout, meaning they already notify all subscribers to a topic (even if they have the same `consumerID`). Examples: Azure Service Bus Topics, Redis Stream (these are already doing what you're describing)
2. Some send a message to one subscriber per each consumer ID. Example: RabbitMQ
3. Some deliver a message to a single subscriber across ALL consumer groups. Example: Azure Service Bus Queues

(All these options are currently contractually-valid because the only thing Dapr promises is that there's "at least once" delivery)

The proposal in this issue will allow making brokers of type 2 above behave as brokers of type 1. However, I don't think it's possible to make brokers of type 3 behave like type 1. Likewise, the next question would be how to make brokers of type 1 behave like type 2 or 3, and that is challenging as well.

**Comment by yaron2** [comment]
Is this about making sure N instances of the same app ID all receive a copy of a message?

**Comment by artursouza** [comment]
> Is this about making sure all instances of the same app ID all receive a copy of a message?

No, this is about a single app receiving the same message via multiple handles and each one can pass/fail independently.

**Comment by yaron2** [comment]
> > Is this about making sure all instances of the same app ID all receive a copy of a message?
> 
> No, this is about a single app receiving the same message via multiple handles and each one can pass/fail independently.

Is handle a route, ie. endpoint?

**Comment by artursouza** [comment]
> Tricky thing with this proposal is that it's very much broker-dependent. Related issue: #5244
> 
> As it is today, we have 3 kinds of brokers:
> 
> 1. Some do fanout, meaning they already notify all subscribers to a topic (even if they have the same `consumerID`). Examples: Azure Service Bus Topics, Redis Stream (these are already doing what you're describing)
> 2. Some send a message to one subscriber per each consumer ID. Example: RabbitMQ
> 3. Some deliver a message to a single subscriber across ALL consumer groups. Example: Azure Service Bus Queues
> 
> (All these options are currently contractually-valid because the only thing Dapr promises is that there's "at least once" delivery)
> 
> The proposal in this issue will allow making brokers of type 2 above behave as brokers of type 1. However, I don't think it's possible to make brokers of type 3 behave like type 1. Likewise, the next question would be how to make brokers of type 1 behave like type 2 or 3, and that is challenging as well.

I think we should solve the broadcast behavior per component-type in a separate issue. It is valid but we need to agree if this variation is something that we should allow in PubSub or should we move some of those to binding?

**Comment by artursouza** [comment]
> > > Is this about making sure all instances of the same app ID all receive a copy of a message?
> > 
> > 
> > No, this is about a single app receiving the same message via multiple handles and each one can pass/fail independently.
> 
> Is handle a route, ie. endpoint?

Correct. See my example JSON. Just different paths in the controller.

**Comment by berndverst** [comment]
consumerId already has special handling in the runtime today - where the runtime might prepend to it.

If I'm understanding correctly, this proposal will prepend "{subscriptionID}-" for the consumergroup / consumerID value for each subscription that specifies this key.

If that's the plan, that's fine by me. We just need to make sure this will be the very last modification to the consumerID key.

**Comment by yaron2** [comment]
> Some do fanout, meaning they already notify all subscribers to a topic (even if they have the same consumerID). Examples: Azure Service Bus Topics, Redis Stream (these are already doing what you're describing)

Redis Streams does not do a fanout, it operates in a classic consumer group fashion.

> Correct. See my example JSON. Just different paths in the controller.

I support adding broadcast, but I'm not sure yet how prefixing the consumerID format would achieve this.

**Comment by ItalyPaleAle** [comment]
Could be a random suffix so the subscriber ID is random per process and subscription.

**Comment by berndverst** [comment]
@yaron2 i think this issue is about being able to subscribe to a single topic multiple times in the same app with different handlers.

This can be done without negative consequences by automatically creating distinct consumer groups.

Some limitations:
- As we know not all PubSub components support multiple consumer groups (they will not deliver to each group)
- Some PubSub components don't support dynamic consumer groups - they need to be created server side

**Comment by yaron2** [comment]
Yeah I meant broadcast in the scope of a single app. Actually I just ran a hacked PoC of this with Kafka and it works as expected. I don't see why we'd need a change in individual components.

> Some PubSub components don't support dynamic consumer groups - they need to be created server side

Can you give an example of one? Even for these components, as long as the consumer group is created with the right format and specified as the right `subscriptionID`, this should work.

**Comment by berndverst** [comment]
> Can you give an example of one? Even for these components, as long as the consumer group is created with the right format and has as the right `subscriptionID`, this should work.

I have a vague memory that EventHub or EventGrid require a consumer group to be created on the server. (Or was this ServiceBus? One of those.)

**Comment by artursouza** [comment]
There is at least one component that manually handles consumerID but I think it is fine if that is done after the runtime appends subscriberID: https://github.com/dapr/components-contrib/blob/master/pubsub/mqtt/mqtt.go

**Comment by yaron2** [comment]
> There is at least one component that manually handles consumerID but I think it is fine if that is done after the runtime appends subscriberID: https://github.com/dapr/components-contrib/blob/master/pubsub/mqtt/mqtt.go
> 
> I updated the proposal that this can be done 100% in runtime.

Great, then I think we can triage it to 1.10 and see if anyone picks this up.

**Comment by berndverst** [comment]
Many components support manually setting the `consumerID` - and that's fine. In runtime there is code that checks for an existing `consumerID` metadata value or it creates a dynamic one. All we need to do is then further prepend to that value.

**Comment by yaron2** [comment]
> Many components support manually setting the `consumerID` - and that's fine. In runtime there is code that checks for an existing `consumerID` metadata value or it creates a dynamic one. All we need to do is then further prepend to that value.

Indeed. we already have constructs that change `consumerID` today like `"{namespace}"`.

**Comment by saber-wang** [comment]
Multiple subscriptions to a single topic in a single application are the most common "problems" we encounter with pubsub using dapr. At the beginning, I repeatedly explained to the developers why it is not supported...

**Comment by CacheCoww** [comment]
I can work on this

**Comment by yaron2** [comment]
> I can work on this

Thanks!

**Comment by ItalyPaleAle** [comment]
Sorry for going back to this, but now that I see the proposed implementation and I have a clearer idea of the issue...

This seems to be implemented by creating multiple subscriptions to the same PubSub component, for the same topic, but with different "consumer groups".

Aside from what I've described above with regards to differences in behavior (e.g. some pubsub components only send a message to 1 subscriber and 1 only), this approach also almost always creates new connections to the pubsub broker, one per each handler, and this is very wasteful.

Couldn't this be implemented by changing the message "router" in the runtime? Once a message is received in the runtime, it would just be routed to multiple endpoints to the app: https://github.com/dapr/dapr/blob/1e636ff158b61186a7699518e758b70b6d44ed9f/pkg/runtime/runtime.go#L878

**Comment by CacheCoww** [comment]
It was my understanding that the goal is to leverage the advantages of the built in pubsub functionality to avoid complicated error handling scenarios as discussed in a related issue: https://github.com/dapr/dapr/issues/3664

It seems like the only way to create multiple consumer groups to allow for fanout within an app is to create multiple components for each required consumer group. The mentioned workaround in the issue above would also create additional components. 

I do not see a less wasteful way to do it with the current implementation unless there is a way to allow for the consumerID to be overwritten when subscribing to a topic so that we do not need to initialize a new instance of the component.

**Comment by ItalyPaleAle** [comment]
I am very concerned with the waste in this case however. A design like this will make it easier for users to add possibly many handlers to each topic, and for each one we'll be creating new TCP connections which are very expensive (and many brokers have a limit on how many connections they accept). It's also very expensive in terms of resources used by Dapr (memory and CPU).

I agree that error handling would be a problem to sort out, but even with the implementation that has multiple subscribers that problem remains. If two subscribers A and B receive a message, and A sends back an ACK, while B doesn't, what happens then? The broker needs to make that decision, and it depends on each broker (and perhaps how they're configured) on what to do.

I would argue that we should rather just deal with the elephant in the room and deal with the problem on the Dapr side. At a high level, there are two ways to deal with this (assume two handlers A and B):

1. Dapr would send an ACK to the broker only if both A and B ACK the message. If at least one of them doesn't, Dapr would NACK the entire message.
2. Dapr would send an ACK when the first one of A and B ACK's it.

We could even offer both behaviors, configurable as an option, although my recommendation would be to set 1 as default.

Of course this could cause a message to be redelivered, but Dapr's promise is "at least once" delivery, not "exactly once". Other situations in Dapr _could_ cause a message to be delivered twice, and apps need to be prepared to handle that case. If A succeeds and B fails, Dapr can also try to re-deliver the message to B before sending a NACK to the broker, so we can handle retries with Resiliency.

This allows us to be consistent across brokers too. If one subscriber ACK's a message and one doesn't, it's unclear what the broker does in that case, and some could retry the delivery to both, some could retry the delivery to only one, or some could not redeliver at all. As I mentioned, there are even brokers who deliver to exactly one subscriber, so having two connections makes only one of them receive the message.

**Comment by onionhammer** [comment]
I don't fully understand the source of the confusion - Most developers would assume it would be effectively the same as if they were two separate apps, just running in the same app. Each unique subscriber endpoint (whether in 2+ endpoints in 1 app or 2+ apps with 1 endpoint each) should have its own "consumer group" with the pubsub broker.

**Comment by waeljammal** [comment]
We require this as well, we have instances where we want to broadcast events that all instances of a different app must receive, having each one be durable and in it's own consumer group guarantees at least once delivery to each of those apps, we use nats for this. We use the stateful sets name as the consumer group so we can have 3 x gateway instances for eg. but each instance has gateway-0/1/2 as the consumer group and this gives us the fan out we require.
